data/src/main/java/com/linkedin/data/codec/entitystream/AbstractDataEncoder.java

```diff
@@ -0,0 +1,337 @@
+    }
+    catch (IOException e)
+    {
+      _writeHandle.error(e);
+    }
+  }
+
+  abstract protected Data.TraverseCallback createTraverseCallback(OutputStream out) throws IOException;
+
+  /**
+   * Pre-process this {@link DataMap} before serializing it.
+   *
+   * <p>This can be overridden by implementations to modify the map before serializing. Implementations may also
+   * choose to directly serialize the map in whatever form they prefer, and return null to indicate that they have
+   * internally handled serialization.</p>
+   */
+  protected DataMap preProcessMap(DataMap dataMap) throws IOException
+  {
+    return dataMap;
+  }
+
+  /**
+   * Pre-process this {@link DataList} before serializing it.
+   *
+   * <p>This can be overridden by implementations to modify the map before serializing. Implementations may also
+   * choose to directly serialize the list in whatever form they prefer, and return null to indicate that they have
+   * internally handled serialization.</p>
+   */
+  protected DataList preProcessList(DataList dataList) throws IOException
+  {
+    return dataList;
+  }
+
+  /**
+   * Create an iterator that will be used by the encoder to iterate over entries of this {@link DataMap} when
+   * serializing.
+   *
+   * <p>This can be overridden by implementations to control the order in which entries are serialized. It is
+   * highly recommended to not modify the iterator or the backing map after this method has been called. Doing so
+   * may result in a {@link java.util.ConcurrentModificationException}</p>
+   */
+  protected Iterator<Map.Entry<String, Object>> createIterator(DataMap dataMap) throws IOException
+  {
+    return dataMap.entrySet().iterator();
+  }
+
+  /**
+   * Create an iterator that will be used by the encoder to iterate over elements of this {@link DataList} when
+   * serializing.
+   *
+   * <p>This can be overridden by implementations to control the order in which elements are serialized. It is
+   * highly recommended to not modify the iterator or the backing list after this method has been called. Doing so
+   * may result in a {@link java.util.ConcurrentModificationException}</p>
+   */
+  protected Iterator<Object> createIterator(DataList dataList) throws IOException
+  {
+    return dataList.iterator();
+  }
+
+  @Override
+  public void onWritePossible()
+  {
+    while (_writeHandle.remaining() > 0)
+    {
+      if (_done)
+      {
+        if (_out.isEmpty())
+        {
+          _writeHandle.done();
+          break;
+        }
+        else
+        {
+          _writeHandle.write(_out.getBytes());
+        }
+      }
+      else if (_out.isFull())
+      {
+        _writeHandle.write(_out.getBytes());
+      }
+      else
+      {
+        try
+        {
+          generate();
+        }
+        catch (Exception e)
+        {
+          _writeHandle.error(e);
+          break;
+        }
+      }
+    }
+  }
+
+  @SuppressWarnings("unchecked")
```

Comment:
I understand that it is moved from "AbstractJacksonDataEncoder"
What runtime unchecked exceptions are expected ?

Reply:
line 242